graphql/src/Mu/GraphQL/Quasi.hs

```diff
@@ -0,0 +1,154 @@
+{-# language DataKinds         #-}
+{-# language OverloadedStrings #-}
+{-# language TemplateHaskell   #-}
+
+module Mu.GraphQL.Quasi where
+
+import           Data.Int
+import qualified Data.Text                     as T
+import           Language.GraphQL.Draft.Parser (parseSchemaDoc)
+import qualified Language.GraphQL.Draft.Syntax as GQL
+import           Language.Haskell.TH
+
+import           Mu.Rpc
+import           Mu.Schema.Definition
+
+-- | Imports an graphql definition written in-line as a 'Schema'.
+graphql :: String -> String -> T.Text -> Q [Dec]
+graphql scName svName schema =
+  case parseSchemaDoc schema of
+    Left e  -> fail ("could not parse graphql spec: " ++ show e)
+    Right p -> graphqlToDecls scName svName p
+
+data Result =
+    GQLSchema Type
+  | GQLService Type
+  deriving (Show, Eq)
+
+graphqlToDecls :: String -> String -> GQL.SchemaDocument -> Q [Dec]
+graphqlToDecls schemaName serviceName (GQL.SchemaDocument types) = do
+  let schemaName'  = mkName schemaName
+      serviceName' = mkName serviceName
+  rs <- traverse (typeToDec schemaName') types
+  let schemaTypes  = [x | GQLSchema  x <- rs]
+      serviceTypes = [x | GQLService x <- rs]
+  schemaDec <- tySynD schemaName' [] (pure $ typesToList schemaTypes)
+  serviceDec <- tySynD serviceName' [] (pure $ typesToList serviceTypes)
+  pure [schemaDec, serviceDec]
+
+typeToDec :: Name -> GQL.TypeDefinition -> Q Result
+typeToDec _ (GQL.TypeDefinitionScalar _)          = fail "scalar types are not supported"
+typeToDec schemaName (GQL.TypeDefinitionObject objs)       = objToDec objs
+  where
+    objToDec :: GQL.ObjectTypeDefinition -> Q Result
+    objToDec (GQL.ObjectTypeDefinition _ nm _ _ flds) =
+      GQLService <$> [t| 'Service $(textToStrLit $ GQL.unName nm) '[]
+          $(typesToList <$> traverse gqlFieldToType flds) |]
+    gqlFieldToType :: GQL.FieldDefinition -> Q Type
+    gqlFieldToType (GQL.FieldDefinition _ fnm args ftyp _) =
+      [t| 'Method $(textToStrLit $ GQL.unName fnm) '[]
+        $(typesToList <$> traverse argToType args)
+        $(retToType ftyp) |]
+    argToType :: GQL.InputValueDefinition -> Q Type
+    argToType (GQL.InputValueDefinition _ aname _ _) =
+      [t| 'ArgSingle 'Nothing '[] ('SchemaRef $(conT schemaName) $(textToStrLit (GQL.unName aname))) |]
+    retToType :: GQL.GType -> Q Type
```

Comment:
For later it's important to know that when you are dealing with the subscription type (which you have to detect from the schema), you should not use `RetSingle` but `RetStream`.